include/seqan3/index/fm_index.hpp

```diff
@@ -0,0 +1,349 @@
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Christopher Pockrandt <christopher.pockrandt AT fu-berlin.de>
+ * \brief Provides the unidirectional seqan3::fm_index.
+ */
+
+#pragma once
+
+#include <sdsl/suffix_trees.hpp>
+
+#include <range/v3/view/reverse.hpp>
+#include <range/v3/view/transform.hpp>
+
+#include <seqan3/index/concept.hpp>
+#include <seqan3/index/fm_index_iterator.hpp>
+#include <seqan3/index/detail/fm_index_iterator.hpp>
+#include <seqan3/index/detail/csa_alphabet_strategy.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+
+namespace seqan3
+{
+
+//!\cond
+// forward declarations
+template <typename index_t>
+class fm_index_iterator;
+
+template <typename index_t>
+class bi_fm_index_iterator;
+//!\endcond
+
+/*!\addtogroup index
+ * \{
+ */
+
+/*!\brief The default FM Index Configuration.
+ * \ingroup fm_index
+ *
+ * \details
+ *
+ * ### Running time / Space consumption
+ *
+ * SAMPLING_RATE = 16
+ * SIGMA: alphabet_size<char_type> where char_type is the seqan3 alphabet type (e.g. dna4 has an alphabet size of 4)
+ * T_BACKWARD_SEARCH: O(log SIGMA)
+ *
+ * \todo Asymptotic space consumption:
+ *
+ */
+struct fm_index_default_traits
+{
+    //!\brief Type of the underlying SDSL index.
+    using sdsl_index_type = sdsl::csa_wt<
+        sdsl::wt_blcd<
+            sdsl::bit_vector,
+            sdsl::rank_support_v<>,
+            sdsl::select_support_scan<>,
+            sdsl::select_support_scan<0>
+        >,
+        16,
+        10000000,
+        sdsl::sa_order_sa_sampling<>,
+        sdsl::isa_sampling<>,
+        sdsl::plain_byte_alphabet
+    >;
+};
+
+/*!\brief The SeqAn FM Index.
+ * \ingroup fm_index
+ * \implements seqan3::fm_index_concept
+ * \tparam text_t The type of the text to be indexed; must satisfy std::ranges::ForwardRange.
+ * \tparam fm_index_traits The traits determining the implementation of the underlying SDSL index;
+                           must satisfy seqan3::fm_index_traits_concept.
+ * \details
+ *
+ * The seqan3::fm_index is a fast and space-efficient string index to search strings and collections of strings.
+ *
+ * ### General information
+ *
+ * Here is a short example on how to build an index and search a pattern using an iterator. Please note that there is a
+ * very powerful search module with a high-level interface \todo seqan3::search that encapsulates the use of iterators.
+ *
+ * \snippet test/snippet/index/fm_index.cpp all
+ *
+ * Here is an example using a collection of strings (e.g. a genome with multiple chromosomes or a protein database):
+ *
+ * Coming soon. Stay tuned!
```

Comment:
We can't index multiple sequences, yet? :scream:  We need that next week!

Reply:
rrahn just wanted to index single sequences :jack_o_lantern: